### DISCUSSION TITLE
can`t use word_timestamps it gives an error.

### DISCUSSION THREAD

**Discussion by furkanguzel161** [openai/whisper#2290] [Q&A]
hello i am trying to use timestamps but i am getting this error.
its a modified version of whisper i modified the latest version from pip to make it compatible with intel gpu, its the same modification leuc has done: https://github.com/openai/whisper/compare/main...leuc:whisper:main
`whisper test.mp3 --model large-v3 --language Turkish --output_format json --device xpu  --word_timestamps True
C:\Users\SlyRebula\miniconda3\envs\whisper\Lib\site-packages\torchvision\io\image.py:13: UserWarning: Failed to load image Python extension: 'Could not find module 'C:\Users\SlyRebula\miniconda3\envs\whisper\Lib\site-packages\torchvision\image.pyd' (or one of its dependencies). Try using the full path with constructor syntax.'If you don't plan on using image functionality from `torchvision.io`, you can ignore this warning. Otherwise, there might be something wrong with your environment. Did you have `libjpeg` or `libpng` installed before building `torchvision` from source?
  warn(
Traceback (most recent call last):
  File "C:\Users\SlyRebula\miniconda3\envs\whisper\Lib\site-packages\whisper\transcribe.py", line 481, in cli
    result = transcribe(model, audio_path, temperature=temperature, **args)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\SlyRebula\miniconda3\envs\whisper\Lib\site-packages\whisper\transcribe.py", line 326, in transcribe
    add_word_timestamps(
  File "C:\Users\SlyRebula\miniconda3\envs\whisper\Lib\site-packages\whisper\timing.py", line 298, in add_word_timestamps
    alignment = find_alignment(model, tokenizer, text_tokens, mel, num_frames, **kwargs)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\SlyRebula\miniconda3\envs\whisper\Lib\site-packages\whisper\timing.py", line 205, in find_alignment
    weights = torch.stack([QKs[_l][_h] for _l, _h in model.alignment_heads.indices().T])
                                                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
RuntimeError: indices expected sparse coordinate tensor layout but got Sparse
Skipping test.mp3 due to RuntimeError: indices expected sparse coordinate tensor layout but got Sparse`

**Comment by Advait251206** [comment]
This error is likely **not caused by `word_timestamps` itself**, but by the modifications made to run Whisper on Intel XPU.

The key part of the traceback is:

```text
RuntimeError: indices expected sparse coordinate tensor layout but got Sparse
```

which occurs here:

```python
weights = torch.stack(
    [QKs[_l][_h] for _l, _h in model.alignment_heads.indices().T]
)
```

During word timestamp generation, Whisper uses `model.alignment_heads`, which is stored as a sparse tensor. The code expects a standard PyTorch sparse COO tensor, but your Intel XPU build appears to be returning a different sparse tensor layout that doesn't support `.indices()` in the same way.

## How to verify

Try running the same command on:

- Official OpenAI Whisper
- CPU (`--device cpu`)
- CUDA (`--device cuda`) if available

Example:

```bash
whisper test.mp3 \
  --model large-v3 \
  --language Turkish \
  --word_timestamps True \
  --device cpu
```

If word timestamps work on CPU but fail on XPU, the issue is almost certainly related to the Intel/XPU modifications rather than Whisper itself.

## Things to check

### 1. Test without XPU

The fastest way to isolate the problem:

```bash
--device cpu
```

If that succeeds, you've confirmed an XPU compatibility issue.

### 2. Check the fork's issue tracker

Since you're using a modified Whisper implementation (similar to the Intel/XPU fork), search for:

- `word_timestamps`
- `alignment_heads`
- `sparse tensor`
- `find_alignment`

Many XPU ports focus on inference and may not fully support Whisper's alignment code used for word-level timestamps.

### 3. Inspect `alignment_heads`

You can add debugging code:

```python
print(model.alignment_heads)
print(model.alignment_heads.layout)
print(type(model.alignment_heads))
```

A working Whisper installation typically expects a sparse COO tensor.

## Possible workaround

If your goal is only word-level timestamps, consider using:

- **faster-whisper**
- **whisperX**

instead of OpenAI Whisper's built-in timestamp alignment.

### whisperX

Pros:

- More accurate word timestamps
- Forced-alignment based
- Often better than native Whisper timestamps

Example:

```bash
pip install whisperx
```

```python
import whisperx

model = whisperx.load_model("large-v3", device="cuda")
result = model.transcribe(audio)

# align words
result_aligned = whisperx.align(...)
```

### faster-whisper

Also supports word timestamps and tends to be more actively maintained for production use.

## Likely root cause

The traceback suggests that the Intel XPU adaptation changed or exposed a different sparse tensor implementation, and Whisper's alignment code assumes:

```python
model.alignment_heads.indices()
```

returns COO-style sparse indices.

Because of that mismatch, the alignment stage fails when `word_timestamps=True`, while normal transcription may still work correctly.

So unless someone has already patched the Intel/XPU fork, this is probably an **XPU sparse tensor compatibility bug in the alignment code**, not a problem with your command syntax.
